pkg/promtail/targets/journaltarget.go

```diff
@@ -0,0 +1,224 @@
+		defaultJournalReaderFunc,
+	)
+}
+
+func journalTargetWithReader(
+	logger log.Logger,
+	handler api.EntryHandler,
+	positions *positions.Positions,
+	jobName string,
+	relabelConfig []*relabel.Config,
+	targetConfig *scrape.JournalTargetConfig,
+	readerFunc journalReaderFunc,
+) (*JournalTarget, error) {
+
+	positionPath := fmt.Sprintf("journal-%s", jobName)
+	position := positions.GetString(positionPath)
+
+	if readerFunc == nil {
+		readerFunc = defaultJournalReaderFunc
+	}
+
+	until := make(chan time.Time)
+	t := &JournalTarget{
+		logger:        logger,
+		handler:       handler,
+		positions:     positions,
+		positionPath:  positionPath,
+		relabelConfig: relabelConfig,
+		labels:        targetConfig.Labels,
+		config:        targetConfig,
+
+		until: until,
+	}
+
+	// Default to system path if not defined. Passing an empty string to
+	// sdjournal is valid but forces reads from the journal to be from
+	// the local machine id only, which contradicts the default behavior
+	// of when a path is specified. To standardize, we manually default the
+	// path here.
+	journalPath := targetConfig.Path
+	if journalPath == "" {
+		journalPath = "/var/log/journal"
+	}
+
+	var err error
+	t.r, err = readerFunc(sdjournal.JournalReaderConfig{
+		Path:      journalPath,
+		Cursor:    position,
+		Since:     targetConfig.Since,
+		Formatter: t.formatter,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "creating journal reader")
+	}
+
+	go func() {
+		err := t.r.Follow(until, ioutil.Discard)
+		if err != nil && err != sdjournal.ErrExpired {
+			level.Error(t.logger).Log("msg", "received error during sdjournal follow", "err", err.Error())
+		}
+	}()
+
+	return t, nil
+}
+
+func (t *JournalTarget) formatter(entry *sdjournal.JournalEntry) (string, error) {
+	ts := time.Unix(0, int64(entry.RealtimeTimestamp)*int64(time.Microsecond))
+
+	msg, ok := entry.Fields["MESSAGE"]
+	if !ok {
+		level.Debug(t.logger).Log("msg", "received journal entry with no MESSAGE field")
+		return journalEmptyStr, nil
+	}
+	entryLabels := makeJournalFields(entry.Fields)
+
+	// Add constant labels
+	for k, v := range t.labels {
+		entryLabels[string(k)] = string(v)
+	}
+
+	processedLabels := relabel.Process(labels.FromMap(entryLabels), t.relabelConfig...)
+
+	processedLabelsMap := processedLabels.Map()
+	labels := make(model.LabelSet, len(processedLabelsMap))
+	for k, v := range processedLabelsMap {
+		if k[0:2] == "__" {
+			continue
+		}
+
+		labels[model.LabelName(k)] = model.LabelValue(v)
+	}
+	if len(labels) == 0 {
+		// No labels, drop journal entry
+		return journalEmptyStr, nil
+	}
+
+	t.positions.PutString(t.positionPath, entry.Cursor)
+	err := t.handler.Handle(labels, ts, msg)
+	return journalEmptyStr, err
+}
+
+// Type returns JournalTargetType.
+func (t *JournalTarget) Type() TargetType {
+	return JournalTargetType
+}
+
+// Ready indicates whether or not the journal is ready to be
+// read from.
+func (t *JournalTarget) Ready() bool {
+	return true
```

Comment:
Any ways to ping journald library to ensure it works without moving the cursor ?

Reply:
Not that we have access to. The closest thing would be checking to see if the journal is at the cursor position we specify but that requires the underlying Journal object, which the JournalReader doesn't expose.

Reply:
ok